07.01-java-tcp-and-udp-programming-(1-of-2)/01-course-material/README.md

```diff
@@ -20,60 +9,85 @@ Based on the original course by O. Liechti and J. Ehrensberger.
-  - [Unicast](#unicast)
-  - [Broadcast](#broadcast)
-  - [Multicast](#multicast)
-- [Messaging patterns](#messaging-patterns)
-- [Service discovery protocols](#service-discovery-protocols)
-- [Practical content](#practical-content)
-  - [Execute the code examples](#execute-the-code-examples)
-  - [Try to emit from multiple emitters at the same time](#try-to-emit-from-multiple-emitters-at-the-same-time)
-  - [Update your application protocol](#update-your-application-protocol)
-  - [Learn to use the debugger](#learn-to-use-the-debugger)
-  - [Explore the Java UDP programming template](#explore-the-java-udp-programming-template)
-  - [Go further](#go-further)
-  - [Compare your solution with the official one](#compare-your-solution-with-the-official-one)
+- [Part 1/2](#part-12)
+  - [Explore the code examples](#explore-the-code-examples)
+  - [The Socket API](#the-socket-api)
+  - [TCP](#tcp)
+  - [UDP](#udp)
+  - [Practical content](#practical-content)
+- [Part 2/2](#part-22)
+  - [Explore the code examples](#explore-the-code-examples-1)
+  - [Unicast, broadcast and multicast](#unicast-broadcast-and-multicast)
+  - [Messaging patterns](#messaging-patterns)
+  - [Service discovery protocols](#service-discovery-protocols)
+  - [Read-eval-print loop (REPL)](#read-eval-print-loop-repl)
+  - [Summary of differences between TCP and UDP](#summary-of-differences-between-tcp-and-udp)
+  - [Practical content](#practical-content-1)
+- [Go further](#go-further)
+  - [Implement the _"Guess the number"_ game](#implement-the-guess-the-number-game)
+  - [Implement the _"Temperature monitoring"_ application](#implement-the-temperature-monitoring-application)
 - [Conclusion](#conclusion)
   - [What did you do and learn?](#what-did-you-do-and-learn)
   - [Test your knowledge](#test-your-knowledge)
 - [Finished? Was it easy? Was it hard?](#finished-was-it-easy-was-it-hard)
 - [Additional resources](#additional-resources)
 - [Solution](#solution)
+- [Optional content](#optional-content)
+  - [Variable length data](#variable-length-data)
 - [Sources](#sources)
 
 ## Objectives
 
-You have seen and experimented with TCP in the previous chapter. You have seen
-that TCP is a connection-oriented protocol. It means that a connection must be
-established before sending data.
+As you have seen in previous courses, applications communicate with each other
+using application protocols.
 
-In this chapter, you will see and experiment with UDP. UDP is mainly used when
-reliability is not required. It is used for streaming, gaming, etc.
+In this course, you will learn how to program your own TCP and UDP clients and
+servers/emitters and receivers in Java.
 
-UDP is sensibly different from TCP and it is important to understand the
-differences between the two protocols.
+This will allow you to create your own network applications, such as a chat
+server, a file server, a web server, a real-time multiplayer game server, etc.
```

Comment:
Suggestion for improving the text :

> As you have seen in previous courses, applications communicate with each other using application protocols.
> 
> In this course, you will learn how to implement such communication by programming your own TCP and UDP clients and servers (or emitters and receivers) in Java.
> 
> This will allow you to create your own network applications, such as a chat
> server, a file server, a web server, a real-time multiplayer game server, etc.
> 
> In other words, now that you understand how application protocols work, you will learn how to put them into practice using TCP and UDP in Java.